File: rubygems/pray-cli/lib/pray/upstream_refresh.rb
# frozen_string_literal: true

module Pray
  module Upstream
    module_function

    def apply_path_upstream_refreshes(project, previous, selected, options)
      user_config = Config.load_user_config
      git_sources = GitSources.prepare_git_sources(
        project.project_root,
        project.manifest.sources,
        previous,
        refresh: options.refresh || options.refresh_source_revisions
      )
      sources = Resolve.source_map(project.manifest.sources)
      changed = false
      project.packages.each do |package|
        next if selected && package.declaration.name != selected
        next unless apply_one_path_upstream(project, package, previous, sources, git_sources, user_config, options)

        changed = true
      end
      changed
    end

    def apply_one_path_upstream(project, package, previous, sources, git_sources, user_config, options)
      new_upstream = package.upstream
      return false unless new_upstream && package.declaration.path

      old_upstream = previous&.package&.find { |entry| entry.name == package.declaration.name }&.upstream
      return false unless old_upstream
      return false if old_upstream.version == new_upstream.version && old_upstream.tree_hash == new_upstream.tree_hash

      old_package, new_package = resolve_upstream_pair(
        project, previous, sources, git_sources, user_config, options, old_upstream, new_upstream
      )
      old_content, local_content, merged = merge_fork_content(package, old_upstream, new_upstream, old_package, new_package)
      write_content_files(package.root, old_content, merged)
      write_refreshed_spec(package, new_package, old_content, local_content, merged)
      true
    end

    def resolve_upstream_pair(project, previous, sources, git_sources, user_config, options, old_upstream, new_upstream)
      old_package = resolve_named(
        project.project_root, sources, git_sources, user_config, previous, options,
        old_upstream.name, "= #{old_upstream.version}", old_upstream.source
      )
      resolved_old = LockedUpstream.new(
        name: old_upstream.name,
        version: old_package.spec.version,
        source: old_upstream.source,
        tree_hash: old_package.tree_hash,
        artifact_hash: old_package.artifact_hash
      )
      ensure_locked_match!(old_upstream, resolved_old)
      new_package = resolve_named(
        project.project_root, sources, git_sources, user_config, previous, options,
        new_upstream.name, "= #{new_upstream.version}", new_upstream.source
      )
      [old_package, new_package]
    end

    def merge_fork_content(package, old_upstream, new_upstream, old_package, new_package)
      old_content = content_file_bytes(old_package.root, old_package.spec)
      new_content = content_file_bytes(new_package.root, new_package.spec)
      local_content = content_file_bytes(package.root, package.spec)
      merged, conflicts = try_merge_content_files(old_content, new_content, local_content)
      unless conflicts.empty?
        raise Error.resolution(
          merge_conflict_message(
            package.declaration.name, old_upstream.name, old_upstream.version, new_upstream.version, conflicts
          )
        )
      end
      [old_content, local_content, merged]
    end

    def write_refreshed_spec(package, new_package, old_content, local_content, merged)
      spec_path = Resolve.find_prayspec_file(package.root)
      updated = PackageSpecRender.fork_spec_after_refresh(
        package.spec,
        new_package.spec,
        File.basename(spec_path),
        clean_replica?(old_content, local_content),
        merged.keys
      )
      Transaction.write_file(spec_path, PackageSpecRender.render_package_spec(updated))
    end

    def resolve_named(project_root, sources, git_sources, user_config, lockfile, options, name, constraint, source)
      Resolve.resolve_package(
        project_root,
        sources,
        git_sources,
        user_config,
        ManifestPackage.new(name: name, constraint: constraint, source: source),
        lockfile,
        options: options
      )
    end

    def content_file_bytes(root, spec)
      paths = content_paths(spec.files)
      if paths.length > ResourceLimits::MAX_ARCHIVE_ENTRIES
        raise Error.integrity("package content exceeds #{ResourceLimits::MAX_ARCHIVE_ENTRIES} files")
      end

      files = {}
      total_bytes = 0
      paths.each do |relative|
        PathSafety.validate_archive_member_path!(relative)
        path = File.join(root, relative)
        raise Error.integrity("package file missing: #{relative}") unless File.file?(path)

        size = File.size(path)
        if size > ResourceLimits::MAX_ARCHIVE_ENTRY_BYTES
          raise Error.integrity("package file exceeds #{ResourceLimits::MAX_ARCHIVE_ENTRY_BYTES} bytes: #{relative}")
        end

        total_bytes += size
        if total_bytes > ResourceLimits::MAX_ARCHIVE_TOTAL_BYTES
          raise Error.integrity("package content exceeds #{ResourceLimits::MAX_ARCHIVE_TOTAL_BYTES} bytes")
        end

        files[relative] = File.binread(path)
      end
      files
    end

    def write_content_files(root, old_content, merged)
      (old_content.keys + merged.keys).each { |path| PathSafety.validate_archive_member_path!(path) }
      old_content.each_key do |path|
        next if merged.key?(path)

        Transaction.remove_file(File.join(root, path))
      end
      merged.each do |relative, bytes|
        Transaction.write_file(File.join(root, relative), bytes)
      end
    end
    private_class_method :apply_one_path_upstream, :resolve_upstream_pair, :merge_fork_content,
      :write_refreshed_spec, :resolve_named
  end
end

File: npmjs/pray-cli/src/resolve/upstream-refresh.ts
import { PrayError } from "../errors.js";
import { prepareGitSources } from "../git/sources.js";
import type { LockedUpstream, Lockfile } from "../lockfile/types.js";
import type { ManifestPackage, ManifestSource } from "../manifest/types.js";
import { findPrayspecFile } from "../package-spec/discovery.js";
import { renderPackageSpec } from "../package-spec/render.js";
import type { PackageSpec } from "../package-spec/types.js";
import { writeProjectFile } from "../transaction/index.js";
import type { ResolveOptions } from "./context.js";
import { resolvePackage } from "./project.js";
import { sourceMap } from "./source-map.js";
import type { ResolvedPackage, ResolvedProject } from "./types.js";
import { ensureLockedUpstreamMatches } from "./upstream.js";
import {
  contentFileBytes,
  localContentForRefresh,
  writeContentFiles,
} from "./upstream-io.js";
import {
  isCleanReplica,
  nextUpstreamConstraint,
  tryMergeContentFiles,
  upstreamMergeConflictMessage,
} from "./upstream-merge.js";

export async function applyPathUpstreamRefreshes(
  project: ResolvedProject,
  previous: Lockfile | undefined,
  selected: string | undefined,
  options: ResolveOptions,
): Promise<boolean> {
  const gitSources = prepareGitSources(
    project.projectRoot,
    project.manifest.sources,
    previous,
    options.refreshSourceRevisions,
    options.offline,
  );
  const sources = sourceMap(project.manifest.sources);
  let changed = false;
  for (const packageEntry of project.packages) {
    if (selected && packageEntry.declaration.name !== selected) {
      continue;
    }
    if (
      await applyOnePathUpstream(
        project,
        packageEntry,
        previous,
        sources,
        gitSources,
        options,
      )
    ) {
      changed = true;
    }
  }
  return changed;
}

async function applyOnePathUpstream(
  project: ResolvedProject,
  packageEntry: ResolvedPackage,
  previous: Lockfile | undefined,
  sources: Map<string, ManifestSource>,
  gitSources: ReturnType<typeof prepareGitSources>,
  options: ResolveOptions,
): Promise<boolean> {
  const newUpstream = packageEntry.upstream;
  if (!newUpstream || !packageEntry.declaration.path) {
    return false;
  }
  const localContent = localContentForRefresh(
    packageEntry.root,
    packageEntry.spec,
  );
  if (localContent.size === 0) {
    return materializeEmptyPathFork(
      project,
      packageEntry,
      previous,
      sources,
      gitSources,
      options,
      newUpstream,
    );
  }
  const oldUpstream = previous?.package.find(
    (entry) => entry.name === packageEntry.declaration.name,
  )?.upstream;
  if (!oldUpstream) {
    return false;
  }
  if (
    oldUpstream.version === newUpstream.version &&
    oldUpstream.tree_hash === newUpstream.tree_hash
  ) {
    return false;
  }
  const oldPackage = await resolveNamed(
    project.projectRoot,
    sources,
    gitSources,
    previous,
    options,
    oldUpstream.name,
    `= ${oldUpstream.version}`,
    oldUpstream.source,
  );
  const resolvedOld: LockedUpstream = {
    name: oldUpstream.name,
    version: oldPackage.spec.version,
    ...(oldUpstream.source ? { source: oldUpstream.source } : {}),
    tree_hash: oldPackage.treeHash,
    artifact_hash: oldPackage.artifactHash,
  };
  ensureLockedUpstreamMatches(oldUpstream, resolvedOld);
  const newPackage = await resolveNamed(
    project.projectRoot,
    sources,
    gitSources,
    previous,
    options,
    newUpstream.name,
    `= ${newUpstream.version}`,
    newUpstream.source,
  );
  const oldContent = contentFileBytes(oldPackage.root, oldPackage.spec);
  const newContent = contentFileBytes(newPackage.root, newPackage.spec);
  const merged = tryMergeContentFiles(oldContent, newContent, localContent);
  if (Array.isArray(merged)) {
    throw PrayError.resolution(
      upstreamMergeConflictMessage(
        packageEntry.declaration.name,
        oldUpstream.name,
        oldUpstream.version,
        newUpstream.version,
        merged,
      ),
    );
  }
  writeContentFiles(packageEntry.root, oldContent, merged);
  writeRefreshedSpec(
    packageEntry,
    newPackage,
    isCleanReplica(oldContent, localContent),
    [...merged.keys()],
  );
  return true;
}

async function materializeEmptyPathFork(
  project: ResolvedProject,
  packageEntry: ResolvedPackage,
  previous: Lockfile | undefined,
  sources: Map<string, ManifestSource>,
  gitSources: ReturnType<typeof prepareGitSources>,
  options: ResolveOptions,
  newUpstream: LockedUpstream,
): Promise<boolean> {
  const newPackage = await resolveNamed(
    project.projectRoot,
    sources,
    gitSources,
    previous,
    options,
    newUpstream.name,
    `= ${newUpstream.version}`,
    newUpstream.source,
  );
  const newContent = contentFileBytes(newPackage.root, newPackage.spec);
  writeContentFiles(packageEntry.root, new Map(), newContent);
  writeRefreshedSpec(packageEntry, newPackage, true, [...newContent.keys()]);
  return true;
}

function writeRefreshedSpec(
  packageEntry: ResolvedPackage,
  newPackage: ResolvedPackage,
  cleanReplica: boolean,
  mergedContentPaths: readonly string[],
): void {
  const specPath = findPrayspecFile(packageEntry.root);
  const updated = forkSpecAfterRefresh(
    packageEntry.spec,
    newPackage.spec,
    cleanReplica,
    mergedContentPaths,
  );
  writeProjectFile(specPath, renderPackageSpec(updated));
}

function forkSpecAfterRefresh(
  local: PackageSpec,
  newUpstream: PackageSpec,
  cleanReplica: boolean,
  mergedContentPaths: readonly string[],
): PackageSpec {
  return {
    ...local,
    files: [...mergedContentPaths],
    exports: cleanReplica ? new Map(newUpstream.exports) : local.exports,
    templates: cleanReplica ? new Map(newUpstream.templates) : local.templates,
    upstream: local.upstream
      ? {
          name: newUpstream.name,
          constraint: nextUpstreamConstraint(
            local.upstream.constraint,
            newUpstream.version,
          ),
        }
      : local.upstream,
  };
}

function resolveNamed(
  projectRoot: string,
  sources: Map<string, ManifestSource>,
  gitSources: ReturnType<typeof prepareGitSources>,
  lockfile: Lockfile | undefined,
  options: ResolveOptions,
  name: string,
  constraint: string,
  source: string | undefined,
): Promise<ResolvedPackage> {
  const declaration: ManifestPackage = {
    name,
    constraint,
    ...(source ? { source } : {}),
    exports: [],
    targets: [],
    features: [],
    groups: [],
    optional: false,
    roles: [],
    bound: false,
  };
  return resolvePackage(
    projectRoot,
    sources,
    gitSources,
    declaration,
    lockfile,
    options,
  );
}
